src/wristpy/processing/calibration.py

```diff
@@ -462,3 +439,191 @@ def _get_sampling_rate(timestamps: pl.Series) -> int:
+                error exceeds the `max_calibration_error` threshold.
+        """
+        logger.debug("Starting calibration.")
+        no_motion_data = _extract_no_motion(
+            acceleration=acceleration,
+            no_motion_threshold=self.no_motion_threshold,
+        )
+        linear_transformation = self._closest_point_fit(no_motion_data=no_motion_data)
+
+        logger.debug(
+            "Calibration successful. Scale: %s, Offset: %s. ",
+            linear_transformation.scale,
+            linear_transformation.offset,
+        )
+
+        calibrated_acceleration = (
+            acceleration.measurements * linear_transformation.scale
+        ) + linear_transformation.offset
+
+        return models.Measurement(
+            measurements=calibrated_acceleration, time=acceleration.time
+        )
+
+    def _closest_point_fit(self, no_motion_data: np.ndarray) -> LinearTransformation:
+        """Find linear transformation parameters that minimzes distance to unit sphere.
+
+        This function implements the scipy optimize.minimze function to find the
+        optimial scale and offset parameters that will minimze the error function,
+        which is defined as the distance of the no_motion acceleration data
+        from the unit sphere (where the unit sphere represents the ideal data points
+        under 1g acceleartion due to Earth's gravity). The initial guess for the
+        scale and offset are chosen as 1/0, we provide constrained bounds for the
+        scale and offset parameters to avoid the the case where the scale can be
+        set to 0 and offset to 1/sqrt(3) (exact unit sphere).
+
+        Args:
+            no_motion_data: The acceleration data during periods of no motion.
+
+        Returns:
+            A LinearTransformation object with scale and offset parameters to be applied
+            to acceleration data for calibration.
+
+        Raises:
+            CalibrationError: If the optimization process fails to converge or if the
+                calibration error is not sufficiently minimized.
+        """
+        start_scale = np.ones(3)
+        start_offset = np.zeros(3)
+
+        def get_distance_to_unit_sphere(params: list[float]) -> float:
+            scale = params[:3]
+            offset = params[3:]
+            distances = np.linalg.norm((no_motion_data * scale) + offset, axis=1) - 1
+            return np.sum(distances**2)
+
+        initial_guess = np.concatenate([start_scale, start_offset])
+        data_range = no_motion_data.max() - no_motion_data.min()
+        scale_bounds = [(0.1, None)] * 3
+        offset_bounds = [(-0.5 * data_range, 0.5 * data_range)] * 3
+        bounds = scale_bounds + offset_bounds
+        result = optimize.minimize(
+            get_distance_to_unit_sphere,
+            initial_guess,
+            options={"maxiter": self.max_iterations},
+            bounds=bounds,
+        )
+        cal_error_end = np.sqrt(result.fun / len(no_motion_data))
```

Comment:
I _think_ this will error if the optimization fails because result.fun might not exist. If that's correct, or if you just don't feel like confirming, bring this into the result.success if-statement.

Reply:
I moved it inside the if statement, since it would only be used if result.success == true. Based on my quick reading of the documentation result.fun would still exist even if success == false (it would be the evaluation at the last iteration)